pkg/addon/controllers/addonconfiguration/addon_configuration_reconciler.go

```diff
@@ -19,16 +22,55 @@ type managedClusterAddonConfigurationReconciler struct {
 func (d *managedClusterAddonConfigurationReconciler) reconcile(
 	ctx context.Context, cma *addonv1alpha1.ClusterManagementAddOn, graph *configurationGraph) (*addonv1alpha1.ClusterManagementAddOn, reconcileState, error) {
 	var errs []error
+	configured := sets.Set[string]{}
 
+	// Update mca config references and configured condition to true for addons to rollout.
 	for _, addon := range graph.getAddonsToUpdate() {
-		mca := d.mergeAddonConfig(addon.mca, addon.desiredConfigs)
-		patcher := patcher.NewPatcher[
-			*addonv1alpha1.ManagedClusterAddOn, addonv1alpha1.ManagedClusterAddOnSpec, addonv1alpha1.ManagedClusterAddOnStatus](
-			d.addonClient.AddonV1alpha1().ManagedClusterAddOns(mca.Namespace))
-		_, err := patcher.PatchStatus(ctx, mca, mca.Status, addon.mca.Status)
+		// update mca config references in status
+		newAddon := d.mergeAddonConfig(addon.mca, addon.desiredConfigs)
+		// update mca configured condition to true
+		d.setCondition(newAddon, metav1.ConditionTrue, "ConfigurationsConfigured", "Configurations configured")
+
+		err := d.patchAddonStatus(ctx, newAddon, addon.mca)
 		if err != nil {
 			errs = append(errs, err)
 		}
+
+		configured.Insert(addon.mca.Namespace)
+	}
+
+	// Update mca configured condition to false for addons with rollout status ToApply.
+	for _, addon := range graph.getAddonsToApply() {
+		if configured.Has(addon.mca.Namespace) {
+			continue
+		}
+		if !meta.IsStatusConditionFalse(addon.mca.Status.Conditions, addonv1alpha1.ManagedClusterAddOnConditionConfigured) {
+			newAddon := addon.mca.DeepCopy()
+			d.setCondition(newAddon, metav1.ConditionFalse, "ConfigurationsNotConfigured", "Configurations updated and not configured yet")
+
+			err := d.patchAddonStatus(ctx, newAddon, addon.mca)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	// Update mca configured condition to true for addons with rollout status Succeeded.
+	// Including:
+	// 1. Addons without configurations, the rollout status is set to success in func setRolloutStatus(),
+	// should set condition to true.
+	// 2. Addons with configurations and already rollout successfully. In upgrade scenario, when the
+	// addon configurations do not change while addon components upgrade, should set condition to true.
+	for _, addon := range graph.getAddonsSucceeded() {
```

Comment:
this is not clear to me, should toUpdate and toApply include all addons need to be set? ToSuccess always comes from toUpdate.

Reply:
For addons that don't define any configs, the rollout status will be set to success in[setRolloutStatus](https://github.com/open-cluster-management-io/ocm/blob/main/pkg/addon/controllers/addonconfiguration/graph.go#L99) directly, and will never appear in the toUpdate or toApply.
Another case is for example, addon defines config A, and in next release, both addon-framework and addon-manager upgrades while it keeps using config A, since the desired config already matches the actual, will not have change to add configured condition in toUpdate or toApply.